### PR TITLE
Create case

### PR DESCRIPTION
**Pull request by cob16**
These are all the objects needed to create a case.
They are all unit tested and a high level feature spec is required when
we come to invoking them in the background job.

The gateway still has some knowledge of a "Complaint" which will be
removed in a future PR.

### CODE REVIEW

File: app/lib/gateway/optics.rb [hunk 1]
```diff
@@ -0,0 +1,41 @@
+module Gateway
+  class Optics
+    class ClientError < StandardError; end
+
+    GET_TOKEN_ENDPOINT = 'https://uat.icasework.com/token?db=hmcts'.freeze
+    ENDPOINT = 'https://uat.icasework.com/createcase?db=hmcts'.freeze
```
Comment (asmega) [inline app/lib/gateway/optics.rb]: not needed for this PR but make these ENV vars

File: app/lib/usecase/optics/create_case.rb [hunk 2]
```diff
@@ -0,0 +1,18 @@
+module Usecase
+  module Optics
+    class CreateCase
+      def initialize(optics_gateway:, presenter:, get_bearer_token:)
+        @optics_gateway = optics_gateway
+        @presenter = presenter
+        @get_bearer_token = get_bearer_token
```
Comment (emileswarts) [inline app/lib/usecase/optics/create_case.rb]: Aware this requests a token on each request.
Caching strategy to be discussed.

File: app/lib/usecase/optics/generate_jwt_token.rb [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+        {
+          iss: api_key,
+          aud: url,
+          iat: Time.now.to_i,
+          exp: Time.now.to_i + ONE_HOUR_IN_SECONDS
+        }
+      end
+
+      attr_reader :url, :api_key, :hmac_secret
+      ONE_HOUR_IN_SECONDS = 3600 * 3600
```
Comment (asmega) [inline app/lib/usecase/optics/generate_jwt_token.rb]: don't think this is quite right 😄

File: app/lib/gateway/optics.rb (hunk 1 above)
Comment (emileswarts) [reply app/lib/gateway/optics.rb]: Yeah, will pull these out

File: app/lib/usecase/optics/generate_jwt_token.rb (hunk 3 above)
Comment (emileswarts) [reply app/lib/usecase/optics/generate_jwt_token.rb]: 😆

File: spec/gateway/optics_spec.rb [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+    let(:expected_body) do
+      URI.encode_www_form(
+        grant_type: 'urn:ietf:params:oauth:grant-type:jwt-bearer',
+        assertion: token
+      )
+    end
+
+    it 'sends a request for a token' do
+      gateway.request_bearer_token(jwt_token: token)
+      expect(WebMock) .to have_requested(:post, 'https://uat.icasework.com/token?db=hmcts').with(
```
Comment (asmega) [inline spec/gateway/optics_spec.rb]: `expect(WebMock) .` there's extra whitespace, not sure how that even works?

File: spec/gateway/optics_spec.rb [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+
+    let(:expected_headers) do
+      {
+        'Authorization' => "Bearer #{bearer_token}",
+        'Content-Type' => 'application/json'
+      }
+    end
+
+    it 'takes a jwt token and body' do
+      gateway.post(body: {}, bearer_token: 'foo')
```
Comment (asmega) [inline spec/gateway/optics_spec.rb]: theres no assertion? or wrap it to asset theres no exception?

File: spec/usecase/optics/create_case_spec.rb [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+        bearer_token: 'some token',
+        body: {}
+      ).once
+    end
+
+    it 'gets a bearer token' do
+      expect(get_bearer_token).to have_received(:execute)
+    end
+
+    xit 'calls present on the presenter' do
```
Comment (asmega) [inline spec/usecase/optics/create_case_spec.rb]: should pass now? or use double?

File: app/lib/usecase/optics/create_case.rb (hunk 2 above)
Comment (asmega) [reply app/lib/usecase/optics/create_case.rb]: i think this is a non-issue for now

File: spec/usecase/optics/create_case_spec.rb (hunk 6 above)
Comment (emileswarts) [reply spec/usecase/optics/create_case_spec.rb]: Tom is currently working on this presenter which will house all the mapping logic.
Rubocop fails if I create a fake to stand in place for this, it requires an instance double.

File: spec/gateway/optics_spec.rb (hunk 5 above)
Comment (emileswarts) [reply spec/gateway/optics_spec.rb]: Good catch, have removed this as it serves no purpose.

File: spec/gateway/optics_spec.rb (hunk 4 above)
Comment (emileswarts) [reply spec/gateway/optics_spec.rb]: Yikes, fixed.  Seemed to have no effect on the test.

File: app/lib/usecase/optics/generate_jwt_token.rb (hunk 3 above)
Comment (emileswarts) [reply app/lib/usecase/optics/generate_jwt_token.rb]: Have removed the hardcoded expiry time.  Not useful to us as this is the default and I doubt we would rely on time to determine when to request a new one.

File: app/lib/gateway/optics.rb [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+
+    def request_bearer_token(jwt_token:)
+      rep = post_jwt(jwt_token)
+      json = JSON.parse rep.body, symbolize_names: true
+      json.fetch(:access_token)
+    end
+
+    def post(body:, bearer_token:)
```
Comment (cob16) [inline app/lib/gateway/optics.rb]: we have a method called `post_jwt` and `post` recommend we rename post to make clear what this methods does (e.g.`post_case`)

Comment (emileswarts) [reply app/lib/gateway/optics.rb]: Not sure we should make this gateway specific to just cases.
This gateway is just a wrapper around the Optics API itself

**Pull request merged**